From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tariq Bashir <tbashir@brightsign.biz>
Date: Fri, 5 Jan 2024 15:16:40 +0000
Subject: OS-16547: patch mime sniffer for html utf8 bom

Change to the mime sniffer to recognise html files with a utf8 bom.
Cherry picked from QtWebEngine change ID: I11c9915a4c23155fc8b8b4e405b69ed48821db5a

diff --git a/net/base/mime_sniffer.cc b/net/base/mime_sniffer.cc
index 826622c7378890d80e338f61601128fa5b7b9f12..e9bddf39aaea09997f92f58fa898f9b4dcc330b4 100644
--- a/net/base/mime_sniffer.cc
+++ b/net/base/mime_sniffer.cc
@@ -371,8 +371,17 @@ static bool SniffForHTML(base::StringPiece content,
 
   // We adopt a strategy similar to that used by Mozilla to sniff HTML tags,
   // but with some modifications to better match the HTML5 spec.
-  base::StringPiece trimmed =
-      base::TrimWhitespaceASCII(content, base::TRIM_LEADING);
+  base::StringPiece trimmed;
+  if (base::StartsWith(content, base::kUtf8ByteOrderMark))
+      trimmed = base::TrimString(content, base::kUtf8ByteOrderMark, base::TrimPositions::TRIM_LEADING);
+  else if (base::StartsWith(content, "\xFE\xFF"))
+      trimmed = base::TrimString(content, "\xFE\xFF", base::TrimPositions::TRIM_LEADING);
+  else if (base::StartsWith(content, "\xFF\xFE"))
+      trimmed = base::TrimString(content, "\xFF\xFE", base::TrimPositions::TRIM_LEADING);
+  else
+      trimmed = content;
+
+  trimmed = base::TrimWhitespaceASCII(trimmed, base::TRIM_LEADING);
 
   // |trimmed| now starts at first non-whitespace character (or is empty).
   return CheckForMagicNumbers(trimmed, kSniffableTags, result);
